### В первом мини-проекте предстоит проанализировать данные о бронировании отелей. 
Описание данных
Имеются следующие переменные:

Hotel – тип отеля (City Hotel или Resort Hotel)

Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным

Lead time – количество дней, прошедших между датой бронирования и датой прибытия

Arrival full date – полная дата прибытия

Arrival date year – год прибытия

Arrival date month – месяц прибытия

Arrival date week number – номер недели прибытия

Arrival date day of month – день прибытия

Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)

Adults – число взрослых

Children – число детей

Babies – число младенцев

Meal – выбранный тип питания

Country – страна происхождения клиента

Reserved room type – тип зарезервированного номера

Assigned room type – тип полученного номера (может отличаться от забронированного)

Customer type – тип бронирования

Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину

Reservation status date – дата обновления статуса

1.	Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/Users/79295/Desktop/GITHUB/Analytical-practice/Pandas/2_bookings.csv',
                encoding = 'windows-1251', sep=';')

In [15]:
df[:7]

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

2.	Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.


In [4]:
df = df.rename(columns={"Is Canceled": "is_canceled",
                   "Lead Time": "lead_time",
                   "arrival full date": "arrival_full_date",
                   "Arrival Date Year": "arrival_date_year",
                   "Arrival Date Month": "arrival_date_month",
                   "Arrival Date Week Number": "arrival_date_week_number",
                   "Arrival Date Day of Month": "arrival_date_day_of_month",
                   "Stays in Weekend nights": "stays_in_weekend_nights",
                   "Stays in week nights": "stays_in_week_nights",
                   "stays total nights": "stays_total_nights",
                   "Adults": "adults",
                   "Hotel": "hotel",
                   "Children": "children",
                   "Babies": "babies",
                   "Meal": "meal",
                   "Country": "country",
                   "Reserved Room Type": "reserved_room_type",
                   "Assigned room type": "assigned_room_type",
                   "customer type": "customer_type",
                   "Reservation Status": "reservation_status",
                   "Reservation status_date": "reservation_status_date"                  
                  }
         )

In [5]:
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')

3.	Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.
(в value_counts() ascending=False по дефолту):

Под успешным, видимо, нужно считать бронирование, которое не было отменено, то есть колонка is_canceled = 0

In [80]:
top_5_countries = df.query("is_canceled == 0")['country'].value_counts()[:5]
top_5_countries

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

4.	На сколько ночей в среднем бронируют отели разных типов?

In [19]:
df.groupby('hotel').agg({"stays_total_nights":"mean"}).round(2)

stays_total_nights
hotel                           
City Hotel                  2.98
Resort Hotel                4.32

5.	Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?

In [36]:
change_room_type = df.query(" assigned_room_type != reserved_room_type ").shape
change_room_type

(14917, 21)

6.	Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов

In [99]:
top_month_2016 = df.query("arrival_date_year == 2016 and is_canceled == 0")['arrival_date_month'].value_counts()
top_month_2016

October      3689
May          3563
September    3372
April        3367
March        3347
August       3238
June         3196
July         3073
November     2818
February     2554
December     2462
January      1691
Name: arrival_date_month, dtype: int64

На какой месяц чаще всего успешно оформляли бронь в 2016?
- October      3689

In [101]:
top_month_2017 = df.query("arrival_date_year == 2017 and is_canceled == 0")['arrival_date_month'].value_counts().head(1)
top_month_2017

May    3551
Name: arrival_date_month, dtype: int64

 Изменился ли самый популярный месяц в 2017?
 - да

In [107]:
city_hotel_cancels_per_years = df.query("hotel == 'City Hotel' and is_canceled == 1") \
                                .groupby('arrival_date_year')['arrival_date_month'] \
                                .value_counts()
city_hotel_cancels_per_years

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

7.	Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?

In [113]:
adults_mean = df.agg({"adults":"mean"})
children_mean = df.agg({"children":"mean"})
babies_mean = df.agg({"babies":"mean"})
print('adults_mean =', adults_mean)
print('children_mean =', children_mean)
print('babies_mean =', babies_mean)

adults_mean = adults    1.856403
dtype: float64
children_mean = children    0.10389
dtype: float64
babies_mean = babies    0.007949
dtype: float64


Переменная adults в нашем датафрейме имеет наибольшее среднее значение.

8.	Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?

In [114]:
df['total_kids'] = df['children'] + df['babies']

In [119]:
df.groupby('hotel').agg({"total_kids":"mean"})

total_kids
hotel                   
City Hotel      0.096311
Resort Hotel    0.142586

In [126]:
df.groupby('hotel').total_kids.mean().round(2)

hotel
City Hotel      0.10
Resort Hotel    0.14
Name: total_kids, dtype: float64

Среднее значение переменной оказалось наибольшим для отелей типа 
- Resort Hotel

9. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate).Укажите, среди какой группы показатель выше. Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитать метрику под названием Churn Rate.Churn rate (отток, коэффициент оттока) – это процент подписчиков (например, на push-уведомления от сайта), которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Иными словами, представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах. В нашем случае Churn Rate - это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов! Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Далее проверьте, среди какой группы пользователей показатель оттока выше.В качестве ответа укажите наибольший % оттока, округленный до 2 знаков после точки (то есть доля 0.24563 будет 24.56% и в ответ пойдёт 24.56)

In [124]:
df['has_kids'] = (df['total_kids'] != 0)

In [125]:
df

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  babies  meal  country  reserved_room_type  \
0                          0  ...       0    BB      PRT                   C   
1                          0  ...       0    BB      PRT                   C   
2                          1  ...       0    BB      GBR                   A   
3                          1  ...       0    BB      GBR                   A   
4                          2  ...       0    BB      GBR                   A   
...                      ...  ...     ...   ...      ...                 ...   
119385                     5  ...       0    BB      BEL                   A   
119386                     5  ...       0    BB      FRA                   E   
119387                     5  ...       0    BB      DEU                   D   
119388                     5  ...       0    BB      GBR                   A   
119389                     7  ...       0    HB      DEU                   A   

       assigned_room_type customer_type reservation_status  \
0                       C     Transient          Check-Out   
1                       C     Transient          Check-Out   
2                       C     Transient          Check-Out   
3                       A     Transient          Check-Out   
4                       A     Transient          Check-Out   
...                   ...           ...                ...   
119385                  A     Transient          Check-Out   
119386                  E     Transient          Check-Out   
119387                  D     Transient          Check-Out   
119388 

In [155]:
with_kids_churn_rate  = round(df.query("has_kids == True and is_canceled == 1")['hotel'].count() \
                        / df.query("has_kids == True")['hotel'].count() * 100, 2)
with_kids_churn_rate

34.95

In [158]:
without_kids_churn_rate = round(df.query("has_kids == False and is_canceled == 1")['hotel'].count() \
                        / df.query("has_kids == False")['hotel'].count() * 100, 2)
without_kids_churn_rate

37.22

Более лаконичный способ решения задачи:

In [160]:
df[['has_kids', 'is_canceled']].value_counts() / df[['has_kids']].value_counts() * 100

has_kids  is_canceled
False     0              62.780998
          1              37.219002
True      0              65.049272
          1              34.950728
dtype: float64

Churn rate выше у посетителей без детей